src/librbd/internal.cc

```diff
@@ -1522,6 +1524,190 @@ bool compare_by_name(const child_info_t& c1, const child_info_t& c2)
+        if (data_pool_id != io_ctx.get_id()) {
+          librados::IoCtx data_io_ctx;
+          r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+          if (r < 0) {
+            lderr(cct) << "rbd: error accessing data pool" << dendl;
+            continue;
+          }
+          data_pool = data_io_ctx.get_pool_name();
+          datapools[data_pool].push_back(entry.id.c_str());
+        } else {
+          datapools[pool_name].push_back(entry.id.c_str());
+        }
+      }
+
+      uint64_t bytes_to_free = 0;
+
+      for (uint8_t i = 0; i < arr.size(); ++i) {
+        json_spirit::mObject obj = arr[i].get_obj();
+        std::string name = obj.find("name")->second.get_str();
+        auto img = datapools.find(name);
+        if (img != datapools.end()) {
+          json_spirit::mObject stats = arr[i].get_obj()["stats"].get_obj();
+          pool_percent_used = stats["percent_used"].get_real();
+          if (pool_percent_used <= threshold) continue;
+
+          bytes_to_free = 0;
+
+          pool_total_bytes = stats["max_avail"].get_uint64() +
+                             stats["bytes_used"].get_uint64();
+
+          auto bytes_threshold = (uint64_t) (pool_total_bytes *
+                                             (pool_percent_used - threshold));
+
+          librbd::Image curr_img;
+          for (const auto &it : img->second) {
+	          r = rbd.open_by_id_read_only(io_ctx, curr_img, it, NULL);
+            if (r < 0) continue;
+
+            uint64_t img_size;
+            curr_img.size(&img_size);
+            r = curr_img.diff_iterate2(nullptr, 0, img_size, false, true,
+                [](uint64_t offset, size_t len, int exists, void *arg) {
+            	    auto *to_free = reinterpret_cast<uint64_t *>(arg);
+                  if (exists)
+                    (*to_free) += len;
+                  return 0;
+                }, &bytes_to_free
+            );
+            if (r < 0) continue;
+            to_be_removed.push_back(it);
+            if (bytes_to_free >= bytes_threshold) break;
+          }
+        }
+      }
+      if (bytes_to_free == 0) {
+        ldout(cct, 10) << "rbd: pool usage is lower than or equal to "
+                       << (threshold * 100)
+                       << "%" << dendl;
+        ldout(cct, 10) << "Nothing to do" << dendl;
+        return 0;
+      }
+    }
+
+    if (expire_ts == 0) {
+	    struct timespec now;
+	    clock_gettime(CLOCK_REALTIME, &now);
+	    expire_ts = now.tv_sec;
+    }
+
+    for (const auto &entry : trash_entries) {
+      if (expire_ts >= entry.deferment_end_time) {
+        to_be_removed.push_back(entry.id.c_str());
+      }
+    }
+
+    uint64_t list_size = to_be_removed.size(), i = 0;
+
+    if (list_size == 0) {
+      ldout(cct, 10) << "rbd: nothing to remove" << dendl;
+    } else {
+      for (const auto &entry_id : to_be_removed) {
+        r = rbd.trash_remove(io_ctx, entry_id, true);
+        if (r < 0) {
```

Comment:
Nit: you can drop all but the ```else``` clause error message since the ```rbd``` CLI will format the error